### DISCUSSION TITLE
Object-oriented box model?

### DISCUSSION THREAD

**Discussion by JamesSample** [NIVANorge/lake_liming_app#17] [General]
@nmarquesin @IvanaH8 

Firstly, I hope I'm not annoying you too much by committing directly to `main`! Since I'm mostly hacking around with the docs and the test code in the `notebooks` folder, I figure it doesn't matter too much. If I do any work on the app itself, I'll be sure to send you a pull request :-) 

There are still some things I need to clarify about the lake model, but I think at least the basic concepts are now solid. I'll update the docs as soon as I get chance. 

In an attempt to pull everything together, last night I had a go at writing an object-oriented version the lake model (as I currently understand it). The code is [here ](https://github.com/NIVANorge/lake_liming_app/blob/main/notebooks/lake_model.py) and a quick example notebook is [here](https://nbviewer.org/github/NIVANorge/lake_liming_app/blob/main/notebooks/oop_box_model.ipynb). It's pretty rough and also possibly a bit rubbish, since I don't write much object-oriented code. However, I suspect an OO implementation of the model might be more flexible/better suited for an application like this. 

Any thoughts? No need to go through it in detail now, but if you think this looks promising I'll keep adding new features to this version. Also happy for feedback on how to implement this better.

Thanks!

**Comment by nmarquesin** [comment]
No problem at all committing directly to main from my part.

I don't have a lot of experience writing OO code, but I'm happy to do it for this project. If you think it is the way forward for this project, let's do it.

Has any new change been done to the lake model since February? I have quite a few hours I can use to on this project during the next two weeks, and I think @IvanaH8 is a bit busy with other projects for the next couple of weeks.

**Comment by JamesSample** [comment]
@nmarquesin Apologies for the delay (again...).

The latest version of the model that I produced is the one [here](https://github.com/NIVANorge/lake_liming_app/blob/main/notebooks/lake_model.py) and it needs merging/integrating with the one you and @IvanaH8 have already implemented based on my initial attempt.

I think most of the class methods are OK, at least for now, but please note the following:

 * `Lake._get_pH0_ApH` will probably be replaced by something else, as I don't think we will use this method for the relationship between Ca and pH. It's OK for now, though. The same applies for `Model._pH_from_delta_Ca`.
 
* `LimeProduct` currently assumes the same instantaneous dissolution and overdosing factors for both Ca and Mg (i.e. init args `id_list` and `od_list`, which are read from the Excel file [here](https://github.com/NIVANorge/lake_liming_app/blob/main/data/lime_products.xlsx)). *I think* this will eventually need generalising to allow different sets of ID and OD values for Ca and Mg (for the same lime type). This will require extending the Excel file and re-writing this function, although probably best to wait until we see some data for Mg and Ca.

* All the plotting functions currently use Matplotlib. They should probably be rewritten to use e.g. Alatir so the plots are interactive in Streamlit (or perhaps take an argument like `output='altair'` to switch between Altair and Matplotlib).

Otherwise, if you can use this model (or some variant of it) as the basis for your web application, that would be great. The notebook [here ](https://github.com/NIVANorge/lake_liming_app/blob/main/notebooks/oop_box_model.ipynb) shows how the model can be used, and I think this covers most/all of the use cases you need for the app.

I also think we can simplify some of the input options so that users only need to supply the following:

- Lake area (km2)
- Lake depth (m)
- Lake residence time (tau; years)
- Flow profile ("kyst", "fjell", "none")
- Lake initial pH
- Lake colour (mgPt/l)

And then choose a lime product from a drop-down list (read from the Excel file). In other words, we can remove the options allowing users to specify initial Ca concentrations for the lake, the inflow and the lake bottom etc.

Good luck! Please let me know if you're unsure about anything and I'll try to help.

Thanks :-)
